services/brig/test/integration/API/OAuth.hs

```diff
@@ -0,0 +1,780 @@
+        ],
+      testGroup "refresh tokens" $
+        [ test m "max active tokens" $ testRefreshTokenMaxActiveTokens o db b,
+          test m "refresh access token - success" $ testRefreshTokenRetrieveAccessToken o b,
+          test m "wrong signature - fail" $ testRefreshTokenWrongSignature o b,
+          test m "no token id - fail" $ testRefreshTokenNoTokenId o b,
+          test m "non-existing id - fail" $ testRefreshTokenNonExistingId o b,
+          test m "wrong client id - fail" $ testRefreshTokenWrongClientId b,
+          test m "wrong client secret - fail" $ testRefreshTokenWrongClientSecret b,
+          test m "wrong grant type - fail" $ testRefreshTokenWrongGrantType b,
+          test m "expired token - fail" $ testRefreshTokenExpiredToken o b,
+          test m "revoked token - fail" $ testRefreshTokenRevokedToken b
+        ],
+      testGroup "oauth applications" $
+        [ test m "list applications with account access" $ testListApplicationsWithAccountAccess b,
+          test m "revoke application account access" $ testRevokeApplicationAccountAccess b
+        ]
+    ]
+
+testRegisterNewOAuthClient :: Brig -> Http ()
+testRegisterNewOAuthClient brig = do
+  let newOAuthClient@(NewOAuthClient expectedAppName expectedUrl) = newOAuthClientRequestBody "E Corp" "https://example.com"
+  cid <- occClientId <$> registerNewOAuthClient brig newOAuthClient
+  uid <- randomId
+  oauthClientInfo <- getOAuthClientInfo brig uid cid
+  liftIO $ do
+    expectedAppName @?= ocName oauthClientInfo
+    expectedUrl @?= ocRedirectUrl oauthClientInfo
+
+testCreateOAuthCodeSuccess :: Brig -> Http ()
+testCreateOAuthCodeSuccess brig = do
+  let newOAuthClient@(NewOAuthClient _ redirectUrl) = newOAuthClientRequestBody "E Corp" "https://example.com"
+  cid <- occClientId <$> registerNewOAuthClient brig newOAuthClient
+  uid <- randomId
+  let scope = OAuthScopes $ Set.fromList [ConversationCreate, ConversationCodeCreate]
+  let state = "foobar"
+  createOAuthCode brig uid (NewOAuthAuthCode cid scope OAuthResponseTypeCode redirectUrl state) !!! do
+    const 302 === statusCode
+    const (Just $ unRedirectUrl redirectUrl ^. pathL) === (fmap getPath . getLocation)
+    const (Just $ ["code", "state"]) === (fmap (fmap fst . getQueryParams) . getLocation)
+    const (Just $ cs state) === (getLocation >=> getQueryParamValue "state")
+
+testCreateOAuthCodeRedirectUrlMismatch :: Brig -> Http ()
+testCreateOAuthCodeRedirectUrlMismatch brig = do
+  let newOAuthClient@(NewOAuthClient _ redirectUrl) = newOAuthClientRequestBody "E Corp" "https://example.com"
+  cid <- occClientId <$> registerNewOAuthClient brig newOAuthClient
+  uid <- randomId
+  let state = "foobar"
+  let differentUrl = mkUrl "https://wire.com"
+  createOAuthCode brig uid (NewOAuthAuthCode cid (OAuthScopes Set.empty) OAuthResponseTypeCode differentUrl state) !!! do
+    const 400 === statusCode
+    const (Just $ "access_denied") === (getLocation >=> getQueryParamValue "error")
```

Comment:
```suggestion
    const (Just $ "access-denied") === (getLocation >=> getQueryParamValue "error")
```

?  unless this is determined by the oauth rfcs, of course.  but we do dashes everywhere else.

Reply:
It is indeed determined by the OAuth RFC [here](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2.1).